File: cpp/lunar.cpp
#include <iostream>
#include <cmath>
#include <ctime>

using namespace std;

double moonPhase(int year, int month, int day) {
    if (month < 3) {
        year--;
        month += 12;
    }
    int a = year / 100;
    int b = a / 4;
    int c = 2 - a + b;
    int e = int(365.25 * (year + 4716));
    int f = int(30.6001 * (month + 1));
    double jd = c + day + e + f - 1524.5;
    double daysSinceNew = jd - 2451549.5;
    double newMoons = daysSinceNew / 29.53;
    double phase = (newMoons - floor(newMoons)) * 29.53;
    return phase;
}

const char* phaseName(double phase) {
    if (phase < 1.845) return "🌑 New Moon";
    if (phase < 5.535) return "🌒 Waxing Crescent";
    if (phase < 9.225) return "🌓 First Quarter";
    if (phase < 12.915) return "🌔 Waxing Gibbous";
    if (phase < 16.605) return "🌕 Full Moon";
    if (phase < 20.295) return "🌖 Waning Gibbous";
    if (phase < 23.985) return "🌗 Last Quarter";
    return "🌘 Waning Crescent";
}

double illumination(double phase) {
    return 50 - 50 * cos(2 * M_PI * phase / 29.53);
}

int main() {
    int year, month, day;
    cout << "Enter date (YYYY MM DD) or 0 for today: ";
    cin >> year;
    if (year == 0) {
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        year = now->tm_year + 1900;
        month = now->tm_mon + 1;
        day = now->tm_mday;
    } else {
        cin >> month >> day;
    }
    double phase = moonPhase(year, month, day);
    cout << "\n📅 " << year << "-" << month << "-" << day << endl;
    cout << phaseName(phase) << " (" << illumination(phase) << "% illuminated)" << endl;
    return 0;
}

File: go/lunar.go
package main

import (
    "fmt"
    "math"
    "time"
)

func moonPhase(year, month, day int) float64 {
    y, m := year, month
    if m < 3 {
        y--
        m += 12
    }
    a := y / 100
    b := a / 4
    c := 2 - a + b
    e := int(365.25 * float64(y+4716))
    f := int(30.6001 * float64(m+1))
    jd := float64(c) + float64(day) + float64(e) + float64(f) - 1524.5
    daysSinceNew := jd - 2451549.5
    newMoons := daysSinceNew / 29.53
    phase := (newMoons - math.Floor(newMoons)) * 29.53
    return phase
}

func phaseName(phase float64) (string, string) {
    switch {
    case phase < 1.845:
        return "🌑 New Moon", "New Moon"
    case phase < 5.535:
        return "🌒 Waxing Crescent", "Waxing Crescent"
    case phase < 9.225:
        return "🌓 First Quarter", "First Quarter"
    case phase < 12.915:
        return "🌔 Waxing Gibbous", "Waxing Gibbous"
    case phase < 16.605:
        return "🌕 Full Moon", "Full Moon"
    case phase < 20.295:
        return "🌖 Waning Gibbous", "Waning Gibbous"
    case phase < 23.985:
        return "🌗 Last Quarter", "Last Quarter"
    default:
        return "🌘 Waning Crescent", "Waning Crescent"
    }
}

func illumination(phase float64) float64 {
    return 50 - 50*math.Cos(2*math.Pi*phase/29.53)
}

func main() {
    var year, month, day int
    fmt.Print("Enter date (YYYY MM DD) or 0 for today: ")
    fmt.Scan(&year)
    if year == 0 {
        now := time.Now()
        year, month, day = now.Year(), int(now.Month()), now.Day()
    } else {
        fmt.Scan(&month, &day)
    }
    phase := moonPhase(year, month, day)
    emoji, name := phaseName(phase)
    illum := illumination(phase)
    fmt.Printf("\n📅 %d-%02d-%02d\n", year, month, day)
    fmt.Printf("%s %s (%.1f%% illuminated)\n", emoji, name, illum)
}
